### PR TITLE
feat(dialog): add beforeClose method

### PR DESCRIPTION
**Pull request by willshowell**
Fixes https://github.com/angular/material2/issues/4647

### CODE REVIEW

File: src/lib/dialog/dialog.spec.ts [hunk 1]
```diff
@@ -141,6 +141,20 @@ describe('MdDialog', () => {
...
+    let dialogRef = dialog.open(PizzaMsg, { viewContainerRef: testViewContainerRef });
+    let beforeCloseCallback = jasmine.createSpy('beforeClose callback');
+
+    dialogRef.beforeClose().subscribe(beforeCloseCallback);
+    dialogRef.close('Bulbasaurus');
+    viewContainerFixture.detectChanges();
+
+    viewContainerFixture.whenStable().then(() => {
+      expect(beforeCloseCallback).toHaveBeenCalledWith('Bulbasaurus');
+      expect(overlayContainerElement.querySelector('md-dialog-container')).toBeNull();
```
Comment (jelbourn) [inline src/lib/dialog/dialog.spec.ts]: This test checks that `beforeClose` occurred, but not _when_ it occurred. I'd restructure to be like
```ts
let beforeCloseInvoked = false;
const beforeCloseHandler = () => {
  expect(overlayContainerElement.querySelector('md-dialog-container')).toBeDefined();
  beforeCloseInvoked = true;
});
...
viewContainerFixture.whenStable().then(() => {
  expect(beforeCloseCallback).toHaveBeenCalledWith('Bulbasaurus');
  expect(overlayContainerElement.querySelector('md-dialog-container')).toBeNull();
});
```

**Comment by willshowell** [comment]
@jelbourn update to the test has been made. Let me know if I misinterpreted your comment!

File: src/lib/dialog/dialog.spec.ts [hunk 2]
```diff
@@ -140,6 +140,25 @@ describe('MdDialog', () => {
     });
   }));
 
+  it('should close a dialog and get back a result before it is closed', async(() => {
+    const dialogRef = dialog.open(PizzaMsg, { viewContainerRef: testViewContainerRef });
```
Comment (jelbourn) [inline src/lib/dialog/dialog.spec.ts]: Omit spaces inside braces

**Comment by kdubious** [comment]
I must be missing something, but how do we cancel the close in beforeClose()?

**Comment by willshowell** [comment]
@kdubious check out https://github.com/angular/material2/issues/4647#issuecomment-302569792

**Comment by kdubious** [comment]
@willshowell, thanks. I read it, but still have a point of confusion.

I'm trying to pass a DTO to a dialog. The dialog has a form (ReactiveForm) that displays the DTO, the user makes their edits, then clicks "Save." At that point, I want the parent component (it maintains a list of the objects being edited in the dialog) to be notified about the current state of the object, so that the parent can manage saving to a remote service. If unsuccessful, the parent would prevent the dialog from closing.

Is there another approach  should consider?

**Comment by willshowell** [comment]
@kdubious I probably should've linked to this thread #3460, particularly https://github.com/angular/material2/pull/3460#issuecomment-319420262. Basically `beforeClose` is simply for getting the close value before the closing animation, but does not allow for synchronous or asynchronous cancellation of the close. If you want to cancel closing of the dialog (via <kbd>esc</kbd> or clicking on the backdrop), then you'll want to use `disableClose` and manually subscribe to the backdrop click and keydown event streams.

It sounds like in your case, you need to run the async validation before the dialog actually closes. If communication with the parent component is the issue, try using a service.

**Comment by kdubious** [comment]
Thanks @willshowell. I ended up going with the service.

**Comment by wiikka** [comment]
@willshowell I really don't see any real use case for beforeClose as it is currently implemented? As requested it would make much more sense if you could cancel the close event (e.g. asking from user if he wants to discard the changes). In that use case it doesn't matter where you get the close event (is it esc, backdrop click or even dialogRef.close()). You should be able to intercept all those and cancel the close if you wish?

**Comment by willshowell** [comment]
@wiikka one of the use cases for beforeClose that was mentioned in the original issue was when you don't care how the dialog was closed, and you don't care to prevent it, but you want to do something with the resulting value without having to wait for the animation to complete.

The conversation in https://github.com/angular/material2/pull/3460, and more specifically https://github.com/angular/material2/pull/3460#issuecomment-319420262,  pretty much covers why an all-encompassing "prevent/cancel close" api wasn't chosen.

**Pull request merged**